### PR TITLE
reduce search complexity for BuiltinScalarFunction

### PR DESCRIPTION
**Pull request by comphead**
# Which issue does this PR close?

Related #6448.

Closes https://github.com/apache/arrow-datafusion/pull/6462

# Rationale for this change

Reduce search complexity when mapping BuiltinScalarFunction and its alias. Currently it is linear search and the place becoming a hotspot. The change is to make the search O(1) using static maps

# What changes are included in this PR?

# Are these changes tested?
Yes

# Are there any user-facing changes?
No

### CODE REVIEW

**Comment by comphead** [comment]
@alamb @2010YOUY01

**Review by 2010YOUY01 — COMMENTED** [review]
This implementation makes it very clear what code to change when adding new functions. Suggested some minor changes, but overall looks good to me, thank you!

File: datafusion/expr/src/built_in_function.rs [hunk 1]
```diff
@@ -429,31 +340,130 @@ impl BuiltinScalarFunction {
     }
 }
```
[the remark this answers is not in the document]
Comment (comphead) [reply datafusion/expr/src/built_in_function.rs:342]: what is that?

File: datafusion/expr/src/built_in_function.rs [hunk 2]
```diff
@@ -429,31 +341,130 @@ impl BuiltinScalarFunction {
...
 
-impl fmt::Display for BuiltinScalarFunction {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        for (func_name, func) in NAME_TO_FUNCTION.iter() {
-            if func == self {
-                return write!(f, "{}", func_name);
-            }
+fn aliases(func: &BuiltinScalarFunction) -> &'static [&'static str] {
```
Comment (2010YOUY01) [inline datafusion/expr/src/built_in_function.rs:344]: Sorry I commented on the wrong line previously @comphead , should be here.

Adding a comment for aliases():
`/// First alias in the array is used to display function names`

File: datafusion/expr/src/built_in_function.rs [hunk 3]
```diff
@@ -429,31 +341,130 @@ impl BuiltinScalarFunction {
...
+        // regex functions
+        BuiltinScalarFunction::RegexpMatch => &["regexp_match"],
+        BuiltinScalarFunction::RegexpReplace => &["regexp_replace"],
+
+        // time/date functions
+        BuiltinScalarFunction::Now => &["now"],
+        BuiltinScalarFunction::CurrentDate => &["current_date"],
+        BuiltinScalarFunction::CurrentTime => &["current_time"],
+        BuiltinScalarFunction::DateBin => &["date_bin"],
+        BuiltinScalarFunction::DateTrunc => &["date_trunc", "datetrunc"],
```
Comment (alamb) [inline datafusion/expr/src/built_in_function.rs:426]: I like how this formulation makes it easier to understand what functions have aliases and which do not 👍

**Review by alamb — APPROVED** [review]
Thanks @comphead  and @2010YOUY01

**Pull request merged**